src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala

```diff
@@ -0,0 +1,89 @@
+
+import scala.concurrent.ExecutionContext
+
+case class ClusterIamSetupException(googleProject: GoogleProject)
+  extends LeoException(s"Error occurred setting up IAM roles in project ${googleProject.value}")
+
+class ClusterHelper(dataprocConfig: DataprocConfig,
+                    gdDAO: GoogleDataprocDAO,
+                    googleComputeDAO: GoogleComputeDAO,
+                    googleIamDAO: GoogleIamDAO)
+                   (implicit val executionContext: ExecutionContext, val system: ActorSystem) extends LazyLogging with Retry {
+
+  implicit val cs = IO.contextShift(executionContext)
+
+  def createClusterIamRoles(googleProject: GoogleProject, serviceAccountInfo: ServiceAccountInfo): IO[Unit] = {
+    updateClusterIamRoles(googleProject, serviceAccountInfo, true)
+  }
+
+  def removeClusterIamRoles(googleProject: GoogleProject, serviceAccountInfo: ServiceAccountInfo): IO[Unit] = {
+    updateClusterIamRoles(googleProject, serviceAccountInfo, false)
+  }
+
+  private def updateClusterIamRoles(googleProject: GoogleProject, serviceAccountInfo: ServiceAccountInfo, create: Boolean): IO[Unit] = {
+    val whenGoogle409: PartialFunction[Throwable, Boolean] = {
+      case t: GoogleJsonResponseException => t.getStatusCode == 409
+      case _ => false
+    }
+
+    val retryIam: (GoogleProject, WorkbenchEmail, Set[String]) => IO[Unit] = (project, email, roles) =>
+      IO.fromFuture[Unit](IO(retryExponentially(whenGoogle409, s"IAM policy change failed for Google project '$project'") { () =>
+        if (create) {
+          googleIamDAO.addIamRolesForUser(project, email, roles)
+        } else {
+          googleIamDAO.removeIamRolesForUser(project, email, roles)
+        }
+      }))
+
+    // Add the Dataproc Worker role in the user's project to the cluster service account, if present.
+    // This is needed to be able to spin up Dataproc clusters using a custom service account.
+    // If the Google Compute default service account is being used, this is not necessary.
+    val dataprocWorkerIO = serviceAccountInfo.clusterServiceAccount.map { email =>
+      retryIam(googleProject, email, Set("roles/dataproc.worker"))
+    } getOrElse IO.unit
+
+    // Add the Compute Image User role in Leo's project to the Google API service account.
+    // This is needed in order to use a custom dataproc VM image.
+    // If a custom image is not being used, this is not necessary.
+    val computeImageUserIO = if (dataprocConfig.customDataprocImage.isDefined) {
+      for {
+        emailOpt <- IO.fromFuture(IO(googleComputeDAO.getGoogleApiServiceAccount(googleProject)))
+        _ <- emailOpt match {
+          case Some(email) => retryIam(dataprocConfig.leoGoogleProject, email, Set("roles/compute.imageUser"))
+          case None => IO.raiseError(ClusterIamSetupException(googleProject))
+        }
+      } yield ()
+    } else IO.unit
+
+    List(dataprocWorkerIO, computeImageUserIO).sequence[IO, Unit].void
```

Comment:
minor, you can do `sequence_` to avoid the extra `void` call..you can also `parSequence_` to make them run in paralell